pkg/operator/ceph/cluster/cleanup.go

```diff
@@ -0,0 +1,185 @@
+		podSpec.Spec.NodeSelector = map[string]string{v1.LabelHostname: hostName}
+		labels := controller.AppLabels(CleanupAppName, cluster.Namespace)
+		labels[CleanupAppName] = "true"
+		job := &batch.Job{
+			ObjectMeta: metav1.ObjectMeta{
+				Name:      jobName,
+				Namespace: cluster.Namespace,
+				Labels:    labels,
+			},
+			Spec: batch.JobSpec{
+				Template: podSpec,
+			},
+		}
+
+		if err := k8sutil.RunReplaceableJob(c.context.Clientset, job, true); err != nil {
+			logger.Error("failed to run cluster clean up job on node %q. %+v", hostName, err)
+		}
+	}
+}
+
+func (c *ClusterController) cleanUpJobContainer(cluster *cephv1.CephCluster) v1.Container {
+	volumeMounts := []v1.VolumeMount{}
+	envVars := []v1.EnvVar{}
+	if cluster.Spec.CleanupPolicy.DeleteDataDirOnHosts != "" && cluster.Spec.DataDirHostPath != "" {
+		hostPathVolumeMount := v1.VolumeMount{Name: volumnName, MountPath: cluster.Spec.DataDirHostPath}
+		volumeMounts = append(volumeMounts, hostPathVolumeMount)
+		envVars = append(envVars, []v1.EnvVar{
+			{Name: dataDirHostPath, Value: cluster.Spec.DataDirHostPath},
+			{Name: namespaceDir, Value: cluster.Namespace},
+		}...)
+	}
+	container := v1.Container{
+		Name:            "host-cleanup",
+		Image:           c.rookImage,
+		SecurityContext: osd.PrivilegedContext(),
+		VolumeMounts:    volumeMounts,
+		Env:             envVars,
+		Args:            []string{"ceph", "clean"}}
+
+	return container
+}
+
+func (c *ClusterController) cleanUpJobTemplateSpec(cluster *cephv1.CephCluster) v1.PodTemplateSpec {
+	volumes := []v1.Volume{}
+	hostPathVolume := v1.Volume{Name: volumnName, VolumeSource: v1.VolumeSource{HostPath: &v1.HostPathVolumeSource{Path: cluster.Spec.DataDirHostPath}}}
+	volumes = append(volumes, hostPathVolume)
+
+	podSpec := v1.PodTemplateSpec{
+		ObjectMeta: metav1.ObjectMeta{
+			Name: CleanupAppName,
+		},
+		Spec: v1.PodSpec{
+			Containers: []v1.Container{
+				c.cleanUpJobContainer(cluster),
+			},
+			Volumes:       volumes,
+			RestartPolicy: v1.RestartPolicyOnFailure,
+		},
+	}
+
+	return podSpec
+}
+
+func (c *ClusterController) waitForCephDaemonCleanUp(cluster *cephv1.CephCluster, retryInterval time.Duration) error {
+	logger.Info("waiting for all the ceph daemons to be cleaned up in the cluster %q", cluster.Namespace)
+	for {
+		cephHosts, err := c.getCephHosts(cluster.Namespace)
+		if err != nil {
+			return errors.Wrap(err, "failed to list ceph daemon nodes")
+		}
+
+		if len(cephHosts) == 0 {
+			logger.Info("all ceph daemons are cleaned up")
+			break
+		}
+
+		logger.Debugf("waiting for ceph daemons in cluster %s to be cleaned up. Retrying in %s",
+			cluster.Namespace, retryInterval.String())
+		<-time.After(retryInterval)
+	}
+
+	return nil
+}
+
+func (c *ClusterController) getCephHosts(namespace string) ([]string, error) {
+	cephPodCount := map[string]int{}
+	cephAppNames := []string{mon.AppName, mgr.AppName, osd.AppName, object.AppName, mds.AppName, rbd.AppName}
```

Comment:
what about ceph-nfs?

Reply:
quick look into the ceph-nfs [code](https://github.com/rook/rook/blob/90c8934d31acbeacd10362001fc5cc6acf273bf8/pkg/operator/ceph/nfs/nfs.go#L70), I see that it does not store any data on the host. Can you please confirm that ?

Reply:
that's correct, forgot about that!